dash-renderer/src/persistence.js

```diff
@@ -0,0 +1,346 @@
+/**
+ * Generalized persistence for component props
+ *
+ * When users input new prop values, they can be stored and reapplied later,
+ * when the component is recreated (changing `Tab` for example) or when the
+ * page is reloaded (depending on `persistence_type`) Storage is tied to
+ * component ID and will not on with components without an ID.
+ *
+ * Renderer handles the mechanics, but components must define a few props:
+ *
+ * - `persistence`: boolean, string, or number. For simple usage, set to `true`
+ *   to enable persistence, omit or set `false` to disable. For more complex
+ *   scenarios, use any truthy value, and change to a *different* truthy value
+ *   when you want the persisted values cleared. (modeled off `uirevision` in)
+ *   plotly.js
+ *   Typically should have no default, but the other persistence props should
+ *   have defaults, so all a user needs to do to enable persistence is set this
+ *   one prop.
+ *
+ * - `persisted_props`: array of prop names allowed to persist. Normally should
+ *   default to the full list of supported props, so they can all be enabled at
+ *   once. The main exception to this is if there's a prop that *can* be
+ *   persisted but most users wouldn't want this.
+ *
+ * - `persistence_type`: one of "local", "session", or "memory", just like
+ *   `dcc.Store`. But the default here should be "local" because the main use
+ *   case is to maintain settings across reloads.
+ *
+ * In addition, if any props require special behavior - particularly for storing
+ * a piece nested inside a larger prop, a component can define a class property
+ * (not a React prop) `persistenceTransforms`. This should be an object:
+ * {
+ *   [propName]: {
+ *     extract: propValue => valueToStore,
+ *     apply: (storedValue, propValue) => newPropValue
+ *   }
```

Comment:
Should we only support one extract from a complex prop or should we support an array of extracts? If so, should we consider a way to expose/select these extracts in the `persistence_props`

Reply:
Right, that was the point of the TODO comment below https://github.com/plotly/dash/pull/903/files#diff-1f9fdd7834ce2c4e31ca7ce6aa890a1fR195

Even if we don't have a current use case, the intent would certainly be clearer if we listed the sub-prop specifically. And then we don't lock ourselves out for later.

So what should this look like? Try to match the structure fully (`"columns[i].name"` or `"columns[].name"`) or just a simple namespace of the outer prop (`"columns.name"`)? I think I'd go for the simpler one; I don't think we want to try to automate extract/apply, there are too many cases to consider. What if we want to extract names keyed by column ID, for example?

(Also note to self: if we do this we also need to only store a change if the extracted piece changed vs. the previous value)

Reply:
>> namespace of the outer prop (`"columns.name"`)

Yes. That's what I had in mind. 

>> I don't think we want to try to automate extract/apply

I don't think so either. Might be possible for simple cases, but that could be added later if we have a clear target.

Reply:
namespaced in https://github.com/plotly/dash/pull/903/commits/78df979b65d60d21e2027a3474931f0281995174